### ISSUE TITLE
Turn camera on/off

### ISSUE THREAD

**Issue by holman** [briis/unifiprotect#159]
So bear with me here.

Ubiquiti, in their infinite wisdom, shipped cameras that don’t actually turn off or on. They’re just always on. And on. And on and on. For such a privacy-invasive product, it’s kind of amazing to me that they haven’t shipped a simple on/off switch so that, say, you could turn off your cameras when you leave the building.

So that’s a bummer, and hopefully they remedy that some time. But I’ve been wondering if we can get around that in the meantime. Basically: can we add privacy zones to “emulate” a camera turning off or on?

It doesn’t look like the concept of privacy zones is known in the codebase yet, and forgive my ignorance if it’s not easily accessed. But it’d be great if we could set some sort of `unifiprotect.emulate_camera_off` mode where we’d draw a privacy zone over the entire feed (and then that might set `set_recording_mode` to `never` as well). And then `emulate_camera_on` would delete the privacy zone and change the recording mode back. (The recording mode might be more of a pain since you’d have to store state on previous modes, but you get the idea.)

Does that make sense? Basically I want to get to the point where I’m effectively blacking out the whole screen from being recorded, disabling saving of video, and we could probably set the mic volume to zero, too. Is it a hack? Sure. But it’d be hugely valuable for me in terms of being able to manually or automatically trigger certain cameras on and off (ie, indoor cameras on when someone’s gone, etc etc).

It’s predicated on the ability to mutate Unifi’s Privacy Zone functionality, though, so that might be the hiccup here.

**Comment by briis** [comment]
Hi Zack,
What is it actually that you want to achieve?
Is it that your camera is not recording?
Don't you want motion to be triggered?

If you set recording_mode to never, the camera does not record anything. 
If you set a privacy zone, then the camera should not trigger on motion in that area and as such it will not record anything if recording mode is different from never.

**Comment by holman** [comment]
Setting `recording_mode` to `never` is one approach, but it still means a camera is live and viewing the room. Even though something might not be actually recording, you still could pull up the app/camera stream and watch in real time.

I’m basically trying to emulate an “off” position for the camera. Not only would the recording mode be “never”, so it wouldn’t record, but anyone watching the camera feed at that point in time would just see a black screen as well.

**Comment by briis** [comment]
OK, got it. Yes the combination of recording_mode never and a privacy zone that fills the whole area, does give you a black screen.
Recording mode is done, so I need to figure out a way to set the privacy_zone (and remove it again). 

I'll look in to it, but no guarantees.

**Comment by briis** [comment]
Hi Zack,
I had some time this morning, so I gave it a shot, and I believe it is working. If you could test it for me it would be great.
If you copy the below files from the *custom_components/unifiprotect* on this Github and place them in the same directory in your HA installation, and then restart HA you should have a new service called `unifiprotect_set_privacy_mode` where you can switch this on or off. (true or false) 

Copy these files:
```yaml
camera.py
const.py
manifest.json
services.yaml
```

**Comment by holman** [comment]
Holy crap that was fast! :star:! Works wonderfully, too. This is really awesome.

The only thing I'm missing is to drop the microphone volume to zero, although now that I'm thinking about it it's probably better to not add that to this same code, since I can just set the mic value directly. So yeah, this is perfect; thanks for the quick turnaround on this! I think there's a bunch of people that will really appreciate having this.

**Comment by briis** [comment]
Thanks. Yes it turned out to be less complicated than I thought, but has still been almost a full day of work.

I thought about both setting recording mode and mic level, but the issue with this is, that these are not plain on/off or true/false values, so I would need somehow to store the ON value when we turn privacy on, and then reset to these values when we turn privacy OFF again. So I decided this would be better handled in a HA script, where you can call the services you want, with tha values you want. 

I still need to update the documentation, and see if I can solve a rare startup bug, before I release, but you can now play with this, and please let me know if you should find bugs, or think of a way to improve.

**Comment by briis** [comment]
Hi Zack,
After a bit mor thinking, I decided to rewrite the service and support set mic level and recording mode in one go.
If you download the same 4 files as before, you can test it out. Now the service should be specified as:

```yaml
entity_id: camera.office
privacy_mode: true
mic_level: 0
recording_mode: never
```

mic_level and recording_mode follow the standards as the independent services. Let me know how this works for you.

**Comment by holman** [comment]
Hey @briis- sorry I couldn't get to this until now. Looking like I'm having problems with the camera entities now- none of them were working, and when I deleted and re-added the integration they didn't look to be getting set up properly:

```
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 199, in _async_setup_platform
    await asyncio.shield(task)
  File "/config/custom_components/unifiprotect/camera.py", line 58, in async_setup_entry
    [
  File "/config/custom_components/unifiprotect/camera.py", line 59, in <listcomp>
    UnifiProtectCamera(
  File "/config/custom_components/unifiprotect/camera.py", line 120, in __init__
    super().__init__(upv_object, protect_data, server_info, camera_id, None)
TypeError: __init__() takes 5 positional arguments but 6 were given
```

**Comment by briis** [comment]
Hi @holman 
Yes sorry, at the time of posting things were working. We are trying to fix another bug right now, and I know that the system is is not in a workable state with the current files. Please go to HACS and just reinstall the latest version (0.6.7), then I hope I can make an official release within the next days.

**Comment by holman** [comment]
No worries! Part of the fun of new code. :) When things are settled I'm happy to dig into this again for ya. Appreciate the work you've done on it.

**Comment by briis** [comment]
Closed in release 0.7.0
